bindings/c/test/unit/unit_tests.cpp

```diff
@@ -0,0 +1,1576 @@
+  p[r.size() - 1]++;
+  return r;
+}
+
+TEST_CASE("strinc") {
+  CHECK(strinc("a").compare("b") == 0);
+  CHECK(strinc("y").compare("z") == 0);
+  CHECK(strinc("!").compare("\"") == 0);
+  CHECK(strinc("*").compare("+") == 0);
+  CHECK(strinc("fdb").compare("fdc") == 0);
+  CHECK(strinc("foundation database 6").compare("foundation database 7") == 0);
+
+  char terminated[] = {'a', 'b', '\xff'};
+  CHECK(strinc(std::string(terminated, 3)).compare("ac") == 0);
+}
+
+// Helper function to add `prefix` to all keys in the given map. Returns a new
+// map.
+std::map<std::string, std::string>
+create_data(std::map<std::string, std::string> &&map) {
+  std::map<std::string, std::string> out;
+  for (const auto & [ key, val ] : map) {
+    out[prefix + key] = val;
+  }
+  return out;
+}
+
+// Clears all data in the database, then inserts the given key value pairs.
+void insert_data(FDBDatabase *db,
+                 const std::map<std::string, std::string> &data) {
+  fdb::Transaction tr(db);
+  auto end_key = strinc(prefix);
+  while (1) {
+    tr.clear_range((const uint8_t *)prefix.c_str(), prefix.size(),
+                   (const uint8_t *)end_key.c_str(), end_key.size());
+    for (const auto & [ key, val ] : data) {
+      tr.set((const uint8_t *)key.c_str(), key.size(),
+             (const uint8_t *)val.c_str(), val.size());
+    }
+
+    fdb::EmptyFuture f1 = tr.commit();
+    fdb_error_t err = wait_future(f1);
+    if (err) {
+      fdb::EmptyFuture f2 = tr.on_error(err);
+      fdb_check(wait_future(f2));
+      continue;
+    }
+    break;
+  }
+}
+
+// Get the value associated with `key_name` from the database. Accepts a list
+// of transaction options to apply (values for options not supported). Returns
+// an optional which will be populated with the result if one was found.
+std::optional<std::string>
+get_value(const uint8_t *key_name, int key_name_length, fdb_bool_t snapshot,
+          std::vector<FDBTransactionOption> options) {
+  fdb::Transaction tr(db);
+  while (1) {
+    for (auto &option : options) {
+      fdb_check(tr.set_option(option, nullptr, 0));
+    }
+    fdb::ValueFuture f1 = tr.get(key_name, key_name_length, snapshot);
+
+    fdb_error_t err = wait_future(f1);
+    if (err) {
+      fdb::EmptyFuture f2 = tr.on_error(err);
+      fdb_check(wait_future(f2));
+      continue;
+    }
+
+    int out_present;
+    char *val;
+    int vallen;
+    fdb_check(f1.get(&out_present, (const uint8_t **)&val, &vallen));
+    return out_present ? std::make_optional(std::string(val, vallen)) : std::nullopt;
+  }
+}
+
+struct GetRangeResult {
+  // List of key-value pairs in the range read.
+  std::vector<std::pair<std::string, std::string>> kvs;
+  // True if values remain in the key range requested.
+  bool more;
+};
+
+// Helper function to get a range of kv pairs. Returns a GetRangeResult struct
+// containing the results of the range read.
+GetRangeResult
+get_range(fdb::Transaction& tr, const uint8_t* begin_key_name,
```

Comment:
Functions that take a transaction argument generally should not have retry logic. The owner of the transaction should handle all retries.